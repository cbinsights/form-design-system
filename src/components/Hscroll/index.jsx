import React, { useState, useRef } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import rafSchd from 'raf-schd';

/**
 * @function getScrollAtributes
 * Returns attributes about user scroll position.
 *
 * @param {Object} e - scroll event from overflow element
 * @param {Number} contentScrollWidth - scrollWidth of inner content element
 */
export const getScrollAtributes = (e, contentScrollWidth) => {
  const { scrollLeft, clientWidth } = e.target;

  /* "end" of scroll is the total width minus the visible area */
  const scrollEnd = contentScrollWidth - clientWidth;

  return {
    scrolled: Boolean(scrollLeft),
    scrollEnd: Boolean(scrollLeft === scrollEnd),
  };
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const Hscroll = ({
  enableFade = true,
  enableGutter = false,
  bgColor = 'white',
  children,
  ...rest
}) => {
  const [isScrolled, setIsScrolled] = useState(false);
  const [isAtScrollEnd, setIsAtScrollEnd] = useState(false);
  const contentEl = useRef(null);

  const containerClassNames = cx('hscroll', `hscroll--bgColor--${bgColor}`, {
    'hscroll--enableFade': enableFade,
    'hscroll--enableGutter': enableGutter,
    'hscroll--scrolled': isScrolled,
    'hscroll--scrollEnd': isAtScrollEnd,
  });

  const onScroll = (e) => {
    const { scrolled, scrollEnd } = getScrollAtributes(e, contentEl.current.scrollWidth);
    rafSchd(setIsScrolled(scrolled));
    rafSchd(setIsAtScrollEnd(scrollEnd));
  };

  return (
    <div
      className={containerClassNames}
      onScroll={enableFade ? onScroll : undefined}
      {...rest}
    >
      <div className="hscroll-overflowEl">
        <div ref={contentEl} className="hscroll-content">
          {children}
        </div>
      </div>
    </div>
  );
};

Hscroll.propTypes = {
  /** adds gradient fades on overflowed sides */
  enableFade: PropTypes.bool,

  /** adds background to scrollbar gutter */
  enableGutter: PropTypes.bool,

  /** match to the background color of the parent */
  bgColor: PropTypes.oneOf(['white', 'haze']),

  /** React children */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
    .isRequired,
};

export default Hscroll;
