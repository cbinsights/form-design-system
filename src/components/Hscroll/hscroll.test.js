import React from 'react';
import { shallow } from 'enzyme';

import Hscroll, { getScrollAtributes } from '.';

const WIDTH_CONTAINER = 400;
const WIDTH_CONTENT = 800;

const renderComponent = (props) => shallow(
  <Hscroll
    {...props}
    style={{ width: `${WIDTH_CONTAINER}px` }}
  >
    <div style={{ width: `${WIDTH_CONTENT}px` }}>long boi</div>
  </Hscroll>
);

describe('Hscroll component', () => {

  it('matches snapshot (default props)', () => {
    const wrapper = renderComponent();
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (disable fade)', () => {
    const wrapper = renderComponent({ enableFade: false });
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (enable gutter)', () => {
    const wrapper = renderComponent({ enableGutter: true });
    expect(wrapper).toMatchSnapshot();
  });

  describe('getScrollAttributes', () => {

    it('returns correct attributes for partially scrolling', () => {
      const MOCK_EVENT = {
        target: {
          scrollLeft: 50,
          clientWidth: WIDTH_CONTAINER
        }
      };
      const { scrolled, scrollEnd } = getScrollAtributes(MOCK_EVENT, WIDTH_CONTENT);

      expect(scrolled).toBe(true);
      expect(scrollEnd).toBe(false);
    });

    it('returns correct attributes for scrolling completely to the end', () => {
      const MOCK_EVENT = {
        target: {
          scrollLeft: WIDTH_CONTAINER,
          clientWidth: WIDTH_CONTAINER
        }
      };
      const { scrolled, scrollEnd } = getScrollAtributes(MOCK_EVENT, WIDTH_CONTENT);

      expect(scrolled).toBe(true);
      expect(scrollEnd).toBe(true);
    });

    it('returns correct attributes for scrolling to the beginning', () => {
      const MOCK_EVENT = {
        target: {
          scrollLeft: 0,
          clientWidth: WIDTH_CONTAINER
        }
      };
      const { scrolled, scrollEnd } = getScrollAtributes(MOCK_EVENT, WIDTH_CONTENT);

      expect(scrolled).toBe(false);
      expect(scrollEnd).toBe(false);
    })

  });

});
