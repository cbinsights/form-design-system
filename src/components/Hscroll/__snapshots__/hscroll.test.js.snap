// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hscroll component matches snapshot (default props) 1`] = `
<div
  className="hscroll hscroll--bgColor--white hscroll--enableFade"
  onScroll={[Function]}
  style={
    Object {
      "width": "400px",
    }
  }
>
  <div
    className="hscroll-overflowEl"
  >
    <div
      className="hscroll-content"
    >
      <div
        style={
          Object {
            "width": "800px",
          }
        }
      >
        long boi
      </div>
    </div>
  </div>
</div>
`;

exports[`Hscroll component matches snapshot (disable fade) 1`] = `
<div
  className="hscroll hscroll--bgColor--white"
  style={
    Object {
      "width": "400px",
    }
  }
>
  <div
    className="hscroll-overflowEl"
  >
    <div
      className="hscroll-content"
    >
      <div
        style={
          Object {
            "width": "800px",
          }
        }
      >
        long boi
      </div>
    </div>
  </div>
</div>
`;

exports[`Hscroll component matches snapshot (enable gutter) 1`] = `
<div
  className="hscroll hscroll--bgColor--white hscroll--enableFade hscroll--enableGutter"
  onScroll={[Function]}
  style={
    Object {
      "width": "400px",
    }
  }
>
  <div
    className="hscroll-overflowEl"
  >
    <div
      className="hscroll-content"
    >
      <div
        style={
          Object {
            "width": "800px",
          }
        }
      >
        long boi
      </div>
    </div>
  </div>
</div>
`;
